enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,375 @@
+with other third-party certificate management systems. But this API introduces inconsistency
+and does not fare well when the user wants to Ingress over directly creating routes.
+
+#### Alternative to new status condition types
+
+An alternative to introducing 3 new condition types, would be to use `metadata.managedFields` 
+on Route to denote to the user which fields are managed by the secret-injector controller and 
+server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate` 
+and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.
+
+This although helps in avoiding addition of new condition types will result in a few inconsistencies,
+
+- When transitioning from using managed certificates to manually managing custom certificates,
+  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update 
+  operation done by the user is only on the `.metadata.annotations`. 
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-reference`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+In scenarios where the user deletes the annotation, it is the controller's 
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e. 
+if the status denotes `RouteCertificateManagedType` and since the required annotation is 
+not present this results in the `.spec.tls` being cleared out and the `.status` updated 
+to `RouteCertificateDefaultType`.
+
+If the user deletes the secret associated with the Route through the `route.openshift.io/tls-secret-reference`
+annotation, the secret-injector controller will delete this annotation automatically and publish 
+an `Event` so that the end user is notified regarding this switch to using the default generated
+certificates since the associated secret is not found. The `.status` is also updated to 
+`RouteCertificateDefaultType`.
+
+
+When transitioning from using a managed certificate to manually managed certificate, the user
+is expected to only delete the annotation. If the previously used `.spec.tls` is not deleted
+by the user the Route will continue to use the same certificate data and `.status` is updated 
+to `RouteCertificateCustomType`.
+
+The new secret-injector controller will not reconcile Route CRs that are owned by the 
+ingress-to-route controller.
+
+### Risks and Mitigations
+
+The new controller that is going to be introduced for this enhancement will have the capability 
+to read secret objects and write the secret data into the route CR. There is precedence 
```

Comment:
is it really a risk if the `route-controller-manager` already has those rights?

Reply:
I have mentioned this here as the Auth team whom we took over cert-manager-operator from were initially told by the Networking team that using secrets for certificates was a no go as this increases the attack surface. But as it turns out,
> There is precedence already for such intrusive permissions required by the ingress-to-route controller.

So it is kind of mitigated.

Reply:
I'm open to removing this as a risk.

Reply:
Using secrets on the edge workloads such as router is no go. Here we are inside the cluster and `route-controller-manager` already has those rights. So, yes I think those rights can be considered as given.

Reply:
Removed content under Risks.